### ISSUE TITLE
[RFC] Provide a util function in distributed_c10d.py for initializing per-machine process groups

### ISSUE THREAD

**Issue by SciPioneer** [pytorch/pytorch#53962]
## 🚀 Feature
Provide a util function for creating per-machine process groups as subgroups of the global communication world and returning the subgroup corresponding to the current machine. 

## Motivation

Some communication optimizations in distributed training may require intra-node communications among the processes within each machine.

For example, since intra-node communication is usually much cheaper than inter-node communication, some parallel SGD algorithms like [hierarchical averaging SGD](https://arxiv.org/pdf/1903.05133.pdf
) run intra-node allreduce more frequently, and inter-node/global allreduce less frequently, rather than always use the same synchronization frequency globally. 

Another example can be [SyncBatchNorm](https://pytorch.org/docs/stable/generated/torch.nn.SyncBatchNorm.html). As the tutorial shows, `torch.nn.SyncBatchNorm` can be applied to subgroups rather than the whole communication world. Such subgroup can often be a per-machine process group in practice.

A few DL algorithms initialize their own per-machine process groups in customized communication libraries.
1. [Detectron2](https://github.com/facebookresearch/detectron2/blob/master/detectron2/utils/comm.py#L14)
2. [SlowFast](https://github.com/facebookresearch/SlowFast/blob/master/slowfast/utils/distributed.py#L268)

It would be nice for PyTorch to provide a util function to initialize per-machine process groups. Particularly, a caveat of [torch.distributed.new_group](https://pytorch.org/docs/master/distributed.html#torch.distributed.new_group
) API is that, even if every rank eventually will only use a single per-machine process group, it must initialize all the per-machine process groups. This can be easily neglected by the users when writing their own implementations. Particularly, #48804 updated SyncBatchNorm tutorial to emphasize this point. (I have also seen some use cases that use the less performant `DataParallel` instead of `DistributedDataParallel` to avoid the error-prone initialization of per-machine process groups.) 

## API

In distributed_c10d.py, we can add a util function, which initializes all the local process groups (e.g., per-machine intra-node process groups), and returns the one corresponding to the current rank.

```python
def init_local_process_groups(local_world_size: int, process_group: torch.distributed.ProcessGroup, tiemout: datatime.timedelta, backend: str) -> torch.distributed.ProcessGroup:
```
where `local_world_size` can be the number of GPUs per machine for this case.

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @agolynski @SciPioneer @H-Huang @mrzzd @cbalioglu

**Comment by mrshenli** [comment]
> Particularly, a caveat of torch.distributed.new_group API is that, even if every rank eventually will only use a single per-machine process group, it must initialize all the per-machine process groups.

Curious, have we considered enhancing `new_group` to remove this limitation instead of introducing an API?

Question: since the name of the proposed API is `init_local_process_groups`, what if I provide `ProcessGroup` instances on different machines to this API? Will it detect that they are on different machines and then error out?

**Comment by SciPioneer** [comment]
> Curious, have we considered enhancing `new_group` to remove this limitation instead of introducing an API?

`new_group` function only takes `ranks` arg as a list of integers, so the API can only specify a single process group at a time, and it won't be able to know how to create subgroups for the rest of ranks, unless a list of integer lists instead of a single integer list is provided. After all, we shouldn't assume that when `new_group` is called, the user always wants to create subgroups of the same size.

However, a 2D list is used as an input, this change in `new_group` may break the backward compatibility, let alone providing a 2D list is not convenient at all.

> Question: since the name of the proposed API is `init_local_process_groups`, what if I provide `ProcessGroup` instances on different machines to this API? Will it detect that they are on different machines and then error out?

I think we can provide some warning for this case, but we shouldn't forbid this kind group division. For example, if a user has 4 2-GPU machines, then the user may want to create two subgroups, where each group has 4 GPUs on 2 machines for more frequent communication, i.e., the rank lists will be [0-3] and [4-7].

To avoid any confusion caused by the term "local", what about naming this function as ` create_new_subgroups`, which mainly takes an arg `subgroup_size`?

**Comment by rohan-varma** [comment]
> To avoid any confusion caused by the term "local", what about naming this function as ` create_new_subgroups`, which mainly takes an arg `subgroup_size`?

If we do this, how will the user be able to specify the particular ranks that are going to be part of the subgroup? Are we still going to always just create groups of the size of the local world (which is what the current proposal is doing)? Otherwise, if `subgroup_size` is say, 3, there could be several valid permutations of groups of size 3.

**Comment by pritamdamania87** [comment]
It seems like instead of doing something like this:

```
for i in n_groups:
  torch.distributed.new_group(...)
```

We want a single API to do this. If so, should we have an API very similar to `new_group` called `new_group_multi` that creates several new_groups and returns the ones that are applicable?

**Comment by SciPioneer** [comment]
> > To avoid any confusion caused by the term "local", what about naming this function as ` create_new_subgroups`, which mainly takes an arg `subgroup_size`?
> 
> If we do this, how will the user be able to specify the particular ranks that are going to be part of the subgroup? Are we still going to always just create groups of the size of the local world (which is what the current proposal is doing)? Otherwise, if `subgroup_size` is say, 3, there could be several valid permutations of groups of size 3.

For this case, we can let all the ranks in the same subgroup increase monotonically for convenience. The proposed API will return the subgroup that contains the current rank, which should be sufficient.

**Comment by SciPioneer** [comment]
> It seems like instead of doing something like this:
> 
> ```
> for i in n_groups:
>   torch.distributed.new_group(...)
> ```

Yes, this util function creates multiple subgroups in a loop, and returns the one which the current rank belongs to.

> We want a single API to do this. If so, should we have an API very similar to `new_group` called `new_group_multi` that creates several new_groups and returns the ones that are applicable?

What about `new_subgroups`? Since one input arg here is the overall group, I think the term "subgroup" sounds better here.

**Comment by zhaojuanmao** [comment]
could we have new_subgroups(input=2DLists)? when input is none, it creates new subgroups for processes in each machine? meanwhile it allows users to specify any subgroups they want?

**Comment by SciPioneer** [comment]
> could we have new_subgroups(input=2DLists)? when input is none, it creates new subgroups for processes in each machine? meanwhile it allows users to specify any subgroups they want?

I think the advantage of this API that takes a 2D list is supporting the creation of subgroups of different sizes. I am not sure how common this case can be. An input of local world size can be more convenient to me.

**Comment by zhaojuanmao** [comment]
> > could we have new_subgroups(input=2DLists)? when input is none, it creates new subgroups for processes in each machine? meanwhile it allows users to specify any subgroups they want?
> 
> I think the advantage of this API that takes a 2D list is supporting the creation of subgroups of different sizes or overlapping subgroups. I am not sure how common these cases can be. It could be also a little more error-prone.
> 
> An input of local world size can be more convenient for creating subgroups of equal size, which is very common.

" An input of local world size can be more convenient for creating subgroups of equal size, which is very common."  ==> that could be done in default when input is None?

**Comment by SciPioneer** [comment]
> > > could we have new_subgroups(input=2DLists)? when input is none, it creates new subgroups for processes in each machine? meanwhile it allows users to specify any subgroups they want?
> > 
> > 
> > I think the advantage of this API that takes a 2D list is supporting the creation of subgroups of different sizes or overlapping subgroups. I am not sure how common these cases can be. It could be also a little more error-prone.
> > An input of local world size can be more convenient for creating subgroups of equal size, which is very common.
> 
> " An input of local world size can be more convenient for creating subgroups of equal size, which is very common." ==> that could be done in default when input is None?

Sounds good to me.

**Comment by SciPioneer** [comment]
Decided to provide 2 APIs separately, one for creating subgroups of equal size, and the other one is specifying each subgroup by a rank list.
